### DISCUSSION TITLE
RmInitAdapter failed when GPU is installed in a PCIe Switch - PLEASE HELP!!!

### DISCUSSION THREAD

**Discussion by ykali-sv** [NVIDIA/open-gpu-kernel-modules#697] [Q&A]
Hi,

I’m developing a pre-Si PCie Switch in emulation environment (FPGA) with a dedicated board that serves as a PCIe bridge connected to the emulator with RTL of the switch.
the GPU is Nvidia RTX 4000 Ada. the emulator is connected to a Supermicro X11DAi-N.
Nvidia Driver version is 550.107.02 on Fedora 37 system.
When the host boots, dmesg shows several issues with the GPU initialization process like: RmInitAdapter: Cannot initialize GSP firmware RM.
when the GPU is connected directly to the host - it works fine!
I’ve attached the dmesg , lspci dump of the GPU and the Nvidia debug log.
I think that maybe because the GPU is slow because its connected an emulator so there are issues with timeout…
Anyone can help me understand why we get these errors ?

[gpu_debug_log.txt](https://github.com/user-attachments/files/16752883/gpu_debug_log.txt)
[gpu_lspci.txt](https://github.com/user-attachments/files/16752885/gpu_lspci.txt)

**Comment by mtijanic** [comment]
Hi there. From the log, it looks like you did manage to read some valid data from PCI already, so it could just be too slow. vbios could be large.

Maybe try something like this to diagnose further?
```diff
diff --git a/src/nvidia/src/kernel/gpu/gsp/arch/turing/kernel_gsp_vbios_tu102.c b/src/nvidia/src/kernel/gpu/gsp/arch/turing/kernel_gsp_vbios_tu102.c
index d179af79..b5c62165 100644
--- a/src/nvidia/src/kernel/gpu/gsp/arch/turing/kernel_gsp_vbios_tu102.c
+++ b/src/nvidia/src/kernel/gpu/gsp/arch/turing/kernel_gsp_vbios_tu102.c
@@ -527,6 +527,8 @@ kgspExtractVbiosFromRom_TU102
 
     biosSize = biosSizeFromRom;
 
+    NV_PRINTF(LEVEL_ERROR, "XXX: vbios size is %u bytes\n", biosSize);
+
     // Copy to system memory and populate pVbiosImg
     {
         NvU32 i;
@@ -542,6 +544,7 @@ kgspExtractVbiosFromRom_TU102
         biosSizeAligned = biosSize & (~0x3);
         for (i = 0; i < biosSizeAligned; i += 4)
         {
+            NV_PRINTF(LEVEL_ERROR, "XXX: Reading vbios word %u out of %u\n", i/4, biosSizeAligned/4);
             pImageDwords[i >> 2] = s_promRead32(pGpu, pciOffset + i);
         }
```

If that's the case, then maybe you can connect the card to a regular PCI bus, extract the vbios and save it to disk, and then patch up that function to return the data from disk instead. However, you'll likely hit other issues as the driver will have to copy much larger data structures eventually (such as uploading of gsp.bin).

**Reply by hrushirajg23** [reply]
@mtijanic 
I'm facing a similar issue:

[ 10.842101] NVRM: s_vbiosPatchInterfaceData: Found pIntFaceHdr entry count: 0, expected 2
[ 10.842107] NVRM: s_vbiosPatchInterfaceData: too few interface entires found for FWSEC cmd 0x15
[ 10.842112] NVRM: s_prepareForFwsec_TU102: Falcon ucode from hs
[ 10.842114] NVRM: s_prepareForFwsec_TU102: failed to prepare interface data for FWSEC cmd 0x15: 0x25
[ 10.842116] NVRM: s_prepareForFwsec_TU102: (note: VBIOS version 94.02.71.40.83)
[ 10.842119] NVRM: nvCheckOkFailedNoLog: Check failed: Invalid data passed [NV_ERR_INVALID_DATA] (0x00000025) returned from kgspPrepareForBootstrap_HAL(pGpu, pKernelGsp, KGSP_BOOT_MODE_NORMAL) @ kernel_gsp.c:3664
[ 10.842171] NVRM: RmInitAdapter: Cannot initialize GSP firmware RM
[ 10.843941] NVRM: GPU 0000:01:00.0: RmInitAdapter failed! (0x62:0x25:2015)
[ 10.845693] NVRM: GPU 0000:01:00.0: rm_init_adapter failed, device minor number 0

Can somebody at-least explain what is the reason behind this "firmware initialization issue" ?
I'm trying to run rtx 3080 on rk3588 sbc .

I added my chipset info to fix the following issue:
[ 6.935721] NVRM: loading NVIDIA UNIX aarch64 Kernel Module 580.95.05 Tue Sep 23 10:16:20 UTC 2025
[ 6.956390] NVRM: Chipset not recognized (vendor ID 0x1d87, device ID 0x3588)
[ 6.956397] The NVIDIA GPU driver for AArch64 has not been qualified on this platform
and therefore it is not recommended or intended for use in any production
environment.

while running proprietary driver I get only:

the chipset error
and
[ 71.348099] NVRM: GPU 0000:01:00.0: RmInitAdapter failed! (0x25:0x65:1623)
[ 71.348111] NVRM: GPU 0000:01:00.0: rm_init_adapter failed, device minor number 0
[ 71.366226] NVRM: GPU at PCI:0000:01:00: GPU-7963b590-e5ce-1751-e3f8-c6d771836da2
[ 71.366233] NVRM: Xid (PCI:0000:01:00): 79, GPU has fallen off the bus.
[ 71.366237] NVRM: GPU 0000:01:00.0: GPU has fallen off the bus.
[ 71.366269] NVRM: A GPU crash dump has been created. If possible, please run
NVRM: nvidia-bug-report.sh as root to collect this data before
NVRM: the NVIDIA kernel module is unloaded.

lspci:
0000:01:00.0 VGA compatible controller: NVIDIA Corporation GA102 [GeForce RTX 3080 Lite Hash Rate] (rev a1)
0000:01:00.1 Audio device: NVIDIA Corporation GA102 High Definition Audio Controller (rev a1)


If you need anymore details like pcie ranges , etc , please let me know. 
Thanks
